spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala

```diff
@@ -2240,7 +2240,15 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
 
   def nullIfWhenPrimitive(expression: Expression): Expression = if (isPrimitive(expression)) {
     val zero = Literal.default(expression.dataType)
-    If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
+    val negZero = Literal.create(-0.0, DoubleType)
+    if (expression.dataType == DoubleType) {
+      If(
+        Or(EqualTo(expression, zero), EqualTo(expression, negZero)),
+        Literal.create(null, expression.dataType),
+        expression)
+    } else {
+      If(EqualTo(expression, zero), Literal.create(null, expression.dataType), expression)
+    }
```

Comment:
I gave another thought, the larger issue is that the following test fails
```
      withParquetTable(Seq((0.0, -0.0)), "tbl") {
        checkSparkAnswerAndOperator("SELECT _1 == _2 FROM tbl")
      }
```
So I think we need to fix EqualTo itself. I opened https://github.com/apache/datafusion/issues/11108

For short term, rather than fixing `nullIfWhenPrimitive`, we can consider fixing `EqualTo` using a similar technique. I.e.
```
 case EqualTo(left, right) =>
    val leftExpr = exprToProtoInternal( If(EqualTo(left, negZero), zero, left), inputs)
    val rightExpr = exprToProtoInternal( If(EqualTo(right, negZero), zero, right), inputs)
```
The above code creates infinity recursive loop, so needs some if-branches to prevent it...

Reply:
for now i have implemented in `nullIfWhenPrimitive`. 

I will explore now how to fix `EqualTo` and subsequently fix this issue.

Reply:
Once `EqualTo` is fixed, we do not need `nullIfWhenPrimitive` workaround

Reply:
Hey @kazuyukitanimura , i tried a lot but I was not able to come up with a solution here. Do you have any pointers for me like how to do this in equalto?

Reply:
@kazuyukitanimura I take it back, ignore the above comment, I have made changes in EqualTo and now it behaves as expected.

i have also added a test case.